modules/animation/src/workspaceanimations.cpp

```diff
@@ -0,0 +1,177 @@
+    std::vector<Animation*> animations;
+    for (auto&& [ind, elem] : util::enumerate(names_)) {
+        if (elem == name) {
+            animations.push_back(&animations_[ind]);
+        }
+    }
+    return animations;
+}
+
+std::string_view WorkspaceAnimations::getName(size_t index) { return names_.at(index); }
+
+Animation& WorkspaceAnimations::add(std::string_view name) {
+    animations_.emplace_back(Animation(&animationManager_));
+    names_.emplace_back(name);
+    auto idx = static_cast<size_t>(animations_.size() - 1);
+    onChanged_.invoke(idx, idx);
+    return animations_.back();
+}
+
+Animation& WorkspaceAnimations::add(std::string_view name, Animation anim) {
+    animations_.emplace_back(anim);
+    names_.emplace_back(name);
+    auto idx = static_cast<size_t>(animations_.size() - 1);
+    onChanged_.invoke(idx, idx);
+    return animations_.back();
+}
+
+Animation& WorkspaceAnimations::insert(size_t index, std::string_view name) {
+    if (index <= getMainAnimationIndex()) {
+        mainAnimationIdx_ = index + 1;
+    }
+    animations_.insert(animations_.begin() + index, Animation(&animationManager_));
+    names_.insert(names_.begin() + index, std::string(name));
+    onChanged_.invoke(index, index);
+    return animations_[index];
+}
+
+void WorkspaceAnimations::erase(size_t from, size_t to) {
+    auto numToRemove = to - from;
+
+    // First ensure that the MainAnimation is valid
+    if (numToRemove == animations_.size()) {
+        add("Animation 1");
+        mainAnimation_.set(animations_.back());
+        mainAnimationIdx_ = 0;  // all before will be removed
+    } else if (getMainAnimationIndex() >= to) {
+        // numToRemove before will be removed, but the animation itself will not change
+        mainAnimationIdx_ -= numToRemove;
+    } else if (getMainAnimationIndex() >= from) {
+        // The selected one will be removed, select next after
+        auto indexBeforeRemoval = to < animations_.size() ? to : from - 1;
+        mainAnimation_.set(animations_[indexBeforeRemoval]);
+        mainAnimationIdx_ = static_cast<size_t>(
+            std::clamp(static_cast<int>(getMainAnimationIndex()) - static_cast<int>(numToRemove), 0,
+                       static_cast<int>(animations_.size()) - 1));
+    }
+
+    animations_.erase(animations_.begin() + from, animations_.begin() + to);
+    names_.erase(names_.begin() + from, names_.begin() + to);
+    onChanged_.invoke(from, to - 1);
+}
+
+void WorkspaceAnimations::setName(size_t index, std::string_view newName) {
+    if (index < names_.size()) {
+        names_[index] = newName;
+        onChanged_.invoke(index, index);
+    }
+}
+
+void WorkspaceAnimations::setMainAnimationIndex(size_t index) {
+    Animation& anim = animations_.at(mainAnimationIdx_);
+    mainAnimationIdx_ = index;
+    mainAnimation_.set(anim);
+}
+
+size_t WorkspaceAnimations::getMainAnimationIndex() const { return mainAnimationIdx_; }
```

Comment:
why is this index exposed, really seems like an implementation detail. is it even needed?

Reply:
It is used by the GUI. I guess it could be made private, but would make the GUI part more complicated.